### PR TITLE
Wallet - Create, Send and Destroy Alias Output

### LINKED ISSUE #436: Wallet - Create and Send Alias Output
## Description

To be able to mint and send native tokens, Wallet module needs to support creating and sending Alias outputs.

### PR DESCRIPTION
**Pull request by matjazv**
# Description of change

- fixed #436 
- fixed deserialization for `sender` and `issuer` feature blocks
- fixed `alias Id` deserialization for alias output
- fully implemented `json_output_deserialize` function 
- fixed alias output `metadata` length to `uint16_t`
- fixed `payload type` serialization for transaction essence 

## Type of change

- Enhancement (a non-breaking change which adds functionality)

## How the change has been tested

All local and tangle unit tests passed.

## Change checklist

Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [x] My code follows the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests using CTest that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/wallet/output_alias.h [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ * @param[in] state_ctrl_addr The receiver address
+ * @param[in] govern_addr The receiver address
+ * @param[out] alias_addr The response of the transfer
+ * @param[out] msg_res The response of the transfer
+ *
+ * @return int 0 on success
+ */
+int wallet_alias_create_send(iota_wallet_t* w, address_t* sender_addr, ed25519_keypair_t* sender_keypair,
+                             uint64_t const send_amount, address_t* state_ctrl_addr, address_t* govern_addr,
+                             address_t* alias_addr, res_send_message_t* msg_res);
```
Comment (oopsmonk) [inline src/wallet/output_alias.h]: the wallet is able to derive the sender address and keypair from BIP44's path inside this method, it also reduces the coding which the user doesn't need to call `get_address_and_keypair`, like in the `examples/wallet/create_alias_output.c` 
maybe rename to `wallet_alias_output_create`?
```suggestion
int wallet_alias_create_send(iota_wallet_t* w, bool sender_change, uint32_t sender_index,
                             uint64_t const send_amount, address_t* state_ctrl_addr, address_t* govern_addr,
                             address_t* alias_addr, res_send_message_t* msg_res);
```

Please add a TODO comment about the optional Feature blocks. let's revisit it when needed.
```
// TODO: the alias output should be able to set optional feature blocks such as Sender/Metadata
```

File: src/wallet/output_alias.h [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+ * @param[in] state_ctrl_addr The state controller address
+ * @param[in] state_ctrl_keypair The state controller private key
+ * @param[in] govern_addr The governor address
+ * @param[out] msg_res The response of the transfer
+ *
+ * @return int 0 on success
+ */
+int wallet_alias_state_transition_send(iota_wallet_t* w, byte_t alias_id[], address_t* state_ctrl_addr,
+                                       ed25519_keypair_t* state_ctrl_keypair, address_t* govern_addr,
+                                       res_send_message_t* msg_res);
```
Comment (oopsmonk) [inline src/wallet/output_alias.h]: the same as `wallet_alias_create_send`, the keypair and address are known by the wallet through the BIP44 path.
maybe rename to `wallet_alias_output_state_transition`?
```suggestion
int wallet_alias_state_transition_send(iota_wallet_t* w, byte_t alias_id[], bool state_ctrl_change, 
                                      uint32_t state_ctrl_index, address_t* govern_addr,
                                       res_send_message_t* msg_res);
```

File: src/wallet/output_alias.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * @param[in] w A wallet instance
+ * @param[in] alias_id The alias identifier
+ * @param[in] govern_keypair The governor private key
+ * @param[in] recv_addr The receiver address
+ * @param[out] msg_res The response of the transfer
+ *
+ * @return int 0 on success
+ */
+int wallet_alias_destroy_send(iota_wallet_t* w, byte_t alias_id[], ed25519_keypair_t* govern_keypair,
+                              address_t* recv_addr, res_send_message_t* msg_res);
```
Comment (oopsmonk) [inline src/wallet/output_alias.h]: the same as `wallet_alias_create_send`, the keypair and address are known by the wallet through the BIP44 path.
maybe rename to `wallet_alias_output_destroy`?
```suggestion
int wallet_alias_destroy_send(iota_wallet_t* w, byte_t alias_id[], bool govern_change, uint32_t govern_index,
                              address_t* recv_addr, res_send_message_t* msg_res);
```

File: src/wallet/output_basic.h [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+ * @param[in] sender_addr The sender address
+ * @param[in] sender_keypair The sender private key
+ * @param[in] send_amount The amount to transfer
+ * @param[in] recv_addr The receiver address
+ * @param[out] msg_res The response of the transfer
+ *
+ * @return int 0 on success
+ */
+int wallet_basic_send(iota_wallet_t* w, address_t* sender_addr, ed25519_keypair_t* sender_keypair,
+                      uint64_t const send_amount, address_t* recv_addr, res_send_message_t* msg_res);
```
Comment (oopsmonk) [inline src/wallet/output_basic.h]: the same as `wallet_alias_create_send`, please revert this change.
fo consistancey, rename to `wallet_basic_output_send`?
```suggestion
int wallet_basic_send(iota_wallet_t* w, bool sender_change, uint32_t sender_index,
                      uint64_t const send_amount, address_t* recv_addr, res_send_message_t* msg_res);
```

File: src/wallet/output_alias.c [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+  }
+
+  // get outputs from the sender address
+  uint64_t total_unspent_amount = 0;
+  unspent_outputs = wallet_get_unspent_basic_outputs(w, sender_addr, sender_keypair, send_amount, tx->essence,
+                                                     &sign_data, &total_unspent_amount);
+  if (!unspent_outputs) {
+    printf("[%s:%d] address does not have any unspent basic outputs\n", __func__, __LINE__);
+    ret = -1;
+    goto end;
```
Comment (oopsmonk) [inline src/wallet/output_alias.c:233]: tx is not freed at the end.
```suggestion
   tx_payload_free(tx);
    ret = -1;
    goto end;
```

File: src/wallet/output_alias.c [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+    printf("[%s:%d] can not create a core message\n", __func__, __LINE__);
+    goto end;
+  }
+
+  // calculate transaction payload ID
+  byte_t payload_id[CRYPTO_BLAKE2B_256_HASH_BYTES] = {0};
+  if (tx_payload_calculate_id(tx, payload_id, sizeof(payload_id)) != 0) {
+    ret = -1;
+    printf("[%s:%d] can not calculate transaction payload ID\n", __func__, __LINE__);
+    goto end;
```
Comment (oopsmonk) [inline src/wallet/output_alias.c]: message is not freed at the end.
```suggestion
    core_message_free(message);
    printf("[%s:%d] can not calculate transaction payload ID\n", __func__, __LINE__);
    goto end;
```

File: src/wallet/output_alias.c [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+
+  cond_blk_free(state);
+  cond_blk_free(governor);
+  cond_blk_list_free(cond_blocks);
+  output_alias_free(alias_output);
+
+  return 0;
+}
+
+int wallet_alias_create_send(iota_wallet_t* w, address_t* sender_addr, ed25519_keypair_t* sender_keypair,
```
Comment (oopsmonk) [inline src/wallet/output_alias.c]: please refactor this method to fix memory leak issues caused by transaction payload and message object.

File: src/wallet/output_alias.c [hunk 8]
```diff
@@ -0,0 +1,423 @@
...
+    printf("[%s:%d] create tx payload failed\n", __func__, __LINE__);
+    ret = -1;
+    goto end;
+  }
+
+  // get unspent alias output
+  output_res = wallet_get_unspent_alias_output(w, alias_id);
+  if (!output_res) {
+    printf("[%s:%d] alias address does not have any unspent alias outputs\n", __func__, __LINE__);
+    ret = -1;
```
Comment (oopsmonk) [inline src/wallet/output_alias.c:340]: the same as `wallet_alias_create_send`, the `tx` memory will be owned by the message, but before message creation, the `tx` has not been freed here, also following exceptions.

File: src/wallet/output_alias.c [hunk 9]
```diff
@@ -0,0 +1,423 @@
...
```
Comment (oopsmonk) [inline src/wallet/output_alias.c:420]: the same as `wallet_alias_create_send`

File: src/wallet/output_basic.c [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+  }
+
+  // get unspent basic outputs from a sender address
+  uint64_t total_unspent_amount = 0;
+  unspent_outputs = wallet_get_unspent_basic_outputs(w, sender_addr, sender_keypair, send_amount, tx->essence,
+                                                     &sign_data, &total_unspent_amount);
+  if (!unspent_outputs) {
+    printf("[%s:%d] address does not have any unspent basic outputs\n", __func__, __LINE__);
+    ret = -1;
+    goto end;
```
Comment (oopsmonk) [inline src/wallet/output_basic.c:240]: the same as `wallet_alias_create_send`,

**Review by oopsmonk — CHANGES_REQUESTED** [review]
Nice work! :+1: 
some suggestions to improve user experience and leak issues.

File: src/wallet/output_alias.h (hunk 1 above)
Comment (matjazv) [reply src/wallet/output_alias.h]: Thank you for all code suggestions and improvements.

File: examples/wallet/send_basic_output.c [hunk 11]
```diff
@@ -9,22 +9,23 @@
...
+#define TEST_COIN_TYPE SLIP44_COIN_TYPE_IOTA
 
 // replace this with your mnemonic string
 static char const* const test_mnemonic =
     "acoustic trophy damage hint search taste love bicycle foster cradle brown govern endless depend situate athlete "
     "pudding blame question genius transfer van random vast";
 static char const* const bech32_receiver =
-    "rms1qrhacyfwlcnzkvzteumekfkrrwks98mpdm37cj4xx3drvmjvnep6xrlkcfw";  // receiver address in bech32 format
-uint32_t const sender_addr_index = 0;                                   // address index of the wallet
+    "atoi1qpswqe4v8z2cdtgc7sfj0hfneqh37lhmjgnth36mfndwcxkjrakcvx42lsn";  // receiver address in bech32 format
```
Comment (oopsmonk) [inline examples/wallet/send_basic_output.c]: it's not related to this PR but could you change this hard-coded bech32 address to the 2nd address index? so this example can work with different network configurations.

**Review by oopsmonk — APPROVED** [review]
lgtm

File: examples/wallet/send_basic_output.c (hunk 11 above)
Comment (matjazv) [reply examples/wallet/send_basic_output.c]: This is a nice improvement. Tnx.
Done.

**Review by rebinmannady93 — APPROVED** [review]

**Pull request merged**